src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -8272,7 +8272,7 @@ private StatementSyntax ParseLocalDeclarationStatement(SyntaxToken awaitKeywordO
             {
                 usingKeyword = CheckFeatureAvailability(usingKeyword, MessageID.IDS_FeatureUsingDeclarations);
             }
-            bool canParseAsLocalFunction = usingKeyword == default;
+            bool canParseAsLocalFunction = usingKeyword == null;
```

Comment:
>null [](start = 59, length = 4)

why did this change?

Reply:
This is one of the two breaking changes described in the PR. Below is the relevant section of the spec. Only `null` and expressions known to have a reference type (ie. not `default`) get to use the object equality in absence of a defined `==` operator.



2.11.7 Reference type equality operators
Every class type C implicitly provides the following predefined reference type equality operators:
  `bool operator ==(C x, C y);`
  `bool operator !=(C x, C y);`
unless predefined equality operators otherwise exist for C (for example, when C is string or System.Delegate).
The operators return the result of comparing the two references for equality or non-equality. operator == returns true if and only if x and y refer to the same instance or are both null, while operator != returns true if and only if operator == with the same operands would return false.

In addition to normal applicability rules (§12.6.4.2), the predefined reference type equality operators require one of the following in order to be applicable:
• Both operands are a value of a type known to be a reference-type or the literal null. Furthermore, an explicit identity or reference conversion (§11.3.5) exists from either operand to the type of the other operand.
• One operand is the literal null, and the other operand is a value of type T where T is a type-parameter that is not known to be a value type, and does not have the value type constraint. 
o If at runtime T is a non-nullable value type, the result of == is false and the result of != is true.
o If at runtime T is a nullable value type, the result is computed from the HasValue property of the operand, as described in (§12.11.10).
o If at runtime T is a reference type, the result is true if the operand is null, and false otherwise.


Unless one of these conditions is true, a binding-time error occurs.
[Note: Notable implications of these rules are:
• It is a binding-time error to use the predefined reference type equality operators to compare two references that are known to be different at binding-time. For example, if the binding-time types of the operands are two class types, and if neither derives from the other, then it would be impossible for the two operands to reference the same object. Thus, the operation is considered a binding-time error.
• The predefined reference type equality operators do not permit value type operands to be compared (except when type parameters are compared to null, which is handled specially).
• Operands of predefined reference type equality operators are never boxed. It would be meaningless to perform such boxing operations, since references to the newly allocated boxed instances would necessarily differ from all other references.
For an operation of the form x == y or x != y, if any applicable user-defined operator == or operator != exists, the operator overload resolution rules (§12.4.5) will select that operator instead of the predefined reference type equality operator. It is always possible to select the predefined reference type equality operator by explicitly casting one or both of the operands to type object. end note]
[Example: The following example checks whether an argument of an unconstrained type parameter type is null.
```
class C<T>
{
      void F(T x) {
           if (x == null) throw new ArgumentNullException();
           …
      }
}
```

The `x == null` construct is permitted even though T could represent a non-nullable value type, and the result is simply defined to be false when T is a non-nullable value type. end example]
For an operation of the form x == y or x != y, if any applicable operator == or operator != exists, the operator overload resolution (§12.4.5) rules will select that operator instead of the predefined reference type equality operator. [Note: It is always possible to select the predefined reference type equality operator by explicitly casting both of the operands to type object. end note]
[Example: The example
```
using System;
class Test
{
      static void Main() {
           string s = "Test";
           string t = string.Copy(s);
           Console.WriteLine(s == t);
           Console.WriteLine((object)s == t);
           Console.WriteLine(s == (object)t);
           Console.WriteLine((object)s == (object)t);
      }
}
```
produces the output
True
False
False
False
The s and t variables refer to two distinct string instances containing the same characters. The first comparison outputs True because the predefined string equality operator (§12.11.8) is selected when both operands are of type string. The remaining comparisons all output False because the overload of operator== in the string type is not applicable when either operand has a binding-time type of object.
Note that the above technique is not meaningful for value types. The example
```
class Test
{
      static void Main() {
           int i = 123;
           int j = 123;
           System.Console.WriteLine((object)i == (object)j);
      }
}
```
outputs False because the casts create references to two separate instances of boxed int values. end example]


---
In reply to: [313659757](https://github.com/dotnet/roslyn/pull/37596#discussion_r313659757) [](ancestors = 313659757)

Reply:
Thanks!

---
In reply to: [313668469](https://github.com/dotnet/roslyn/pull/37596#discussion_r313668469) [](ancestors = 313668469,313659757)